src/qutip_qip/vqa.py

```diff
@@ -0,0 +1,852 @@
+            )
+        if isinstance(self.operator, ParameterizedHamiltonian):
+            return self.get_unitary_frechet_derivative(angles)
+        else:
+            if len(angles) != 1:
+                raise ValueError(
+                    "Expected a single angle for non-"
+                    "ParameterizedHamiltonian instance."
+                )
+            return self.get_unitary(angles) * -1j * self.operator
+
+    def get_unitary_frechet_derivative(self, angles, term_index=0):
+        """
+        Compute the derivative of the block's unitary with respect
+        to a free parameter, using the Frechet derivative of
+        the exponential function.
+
+        Parameters
+        ----------
+        angles: list of float
+            the free parameters to take derivatives with respect to
+        term_index: int, optional
+            Index of Parameterized Hamiltonian term that specifies the matrix
+            direction in which to take the derivative.
+
+        Returns
+        -------
+        derivative: float
+        """
+        if self.is_unitary or self.is_native_gate:
+            raise ValueError(
+                "Can only take frechet derivative of block "
+                "specified by Hamiltonians"
+            )
+        if isinstance(self.operator, ParameterizedHamiltonian):
+            arg = -1j * self.operator.get_hamiltonian(angles)
+            direction = -1j * self.operator.p_terms[term_index]
+        else:
+            if len(angles) != 1:
+                raise ValueError(
+                    "Expected a single angle for non-"
+                    "ParameterizedHamiltonian instance."
+                )
+            arg = -1j * self.operator * angles[0]
+            direction = -1j * self.operator
+        return Qobj(
+            expm_frechet(arg, direction, compute_expm=False),
+            dims=direction.dims,
+        )
+
+
+class OptimizationResult:
+    """
+    Class for results of VQA optimization loop.
+
+    Parameters
+    ----------
+    res: scipy results instance
+    final_state: Qobj
+        Final state of the circuit after optimization.
+    """
+
+    def __init__(self, res, final_state):
+        self.res = res
+        self.angles = res.x
+        self.min_cost = res.fun
+        self.nfev = res.nfev
+        self.final_state = final_state
+
+    def _highest_prob_bitstring(self, state):
+        """
+        Return the bitstring associated with the
+        highest probability amplitude measurement state.
+        """
+        n_qubits = int(np.log2(state.shape[0]))
+        index = np.argmax(abs(state))
+        return format(index, f"0{n_qubits}b")
+
+    def get_top_bitstring(self):
+        """
+        Return the bitstring associated with the highest probability
+        measurement outcome
+
+        Returns
+        -------
+        bitstring: str
+            bitstring in the form |x_0x_1...x_n> where each x_i is '0' or '1'
```

Comment:
Better warp `|x_0x_1...x_n>` with the math mode, ``:math:`|x_0x_1...x_n>` ``. Underscores have special meanings in rst files, so the building doc test will complain.